modulus/models/afno/modembed.py

```diff
@@ -0,0 +1,118 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Type
+
+import torch
+from torch import Tensor, nn
+
+
+class PositionalEmbedding(nn.Module):
+    """
+    A module for generating positional embeddings based on timesteps.
+
+    Parameters:
+    -----------
+    num_channels : int
+        Number of channels for the embedding.
+    """
+
+    def __init__(self, num_channels: int):
+        super().__init__()
+        self.num_channels = num_channels
+
+        freqs = torch.pi * torch.arange(
+            start=1, end=self.num_channels // 2 + 1, dtype=torch.float32
+        )
+        self.register_buffer("freqs", freqs)
+
+    def forward(self, x: Tensor):
+        x = x.view(-1).outer(self.freqs.to(x.dtype))
+        x = torch.cat([x.cos(), x.sin()], dim=1)
+        return x
+
+
+class OneHotEmbedding(nn.Module):
+    """
+    A module for generating one-hot embeddings based on timesteps.
+
+    Parameters:
+    -----------
+    num_channels : int
+        Number of channels for the embedding.
+    """
+
+    def __init__(self, num_channels: int):
+        super().__init__()
+        self.num_channels = num_channels
+        ind = torch.arange(num_channels)
+        ind = ind.view(1, len(ind))
+        self.register_buffer("indices", ind)
+
+    def forward(self, t: Tensor):
+        ind = t * (self.num_channels - 1)
+        return torch.clamp(1 - torch.abs(ind - self.indices), min=0)
+
+
+class ModEmbedNet(nn.Module):
+    """
+    A network that generates a timestep embedding and processes it with an MLP.
+
+    Parameters:
+    -----------
+    max_time : float, optional
+        Maximum input time. The inputs to `forward` is should be in the range [0, max_time].
+    dim : int, optional
+        The dimensionality of the time embedding.
+    depth : int, optional
+        The number of layers in the MLP.
+    activation_fn:
+        The activation function, default GELU.
+    method : str, optional
+        The embedding method. Either "sinusoidal" (default) or "onehot".
+    """
+
+    def __init__(
+        self,
+        max_time: float = 1.0,
+        dim: int = 64,
+        depth: int = 1,
+        activation_fn: Type[nn.Module] = nn.GELU,
+        method: str = "sinusoidal",
+    ):
+        super().__init__()
+        self.max_time = max_time
+        self.method = method
+        if method == "onehot":
+            self.onehot_embed = OneHotEmbedding(dim)
```

Comment:
Can you make this and the following line `self.emb = ...` and get rid of the conditional in the forward?

Reply:
Good idea, but wouldn't this break compatibility with existing checkpoints (since AFAIK layer names need to match those in the checkpoint file)?

Reply:
Yeah this would require some manual stitching of the previous checkpoint.

I'd be for making the update, I think the fix for the checkpoint should be a rename of the weight key.
```python
checkpoint = torch.load(load_name)
for key, value in checkpoint['model'].items():
    if key == "modafno.something.something.onehot_embed":
           checkpoint['model']['"modafno.something.something.embed"] = value
```

I think this would be alright (I had to do this before for I think the original AFNO checkpoint :P ). But naturally we need to make sure we have an original implementation to make sure we outputs are exact. Can help if needed.